pkg/lib/registry/registry.go

```diff
@@ -402,3 +411,80 @@ func checkForBundlePaths(querier registry.GRPCQuery, bundlePaths []string) ([]st
 	}
 	return found, missing, nil
 }
+
+// replaces mode selects highest version as channel head and
+// prunes any bundles in the upgrade chain after the channel head.
+// check for the presence of all bundles after a replaces-mode add.
+func checkForBundles(ctx context.Context, q *sqlite.SQLQuerier, g registry.GraphLoader, bundles map[image.Reference]string) error {
+	if len(bundles) == 0 {
+		return nil
+	}
+
+	foundEntries := map[string]map[string]map[registry.BundleKey]bool{}
+	var errs []error
+	for to, from := range bundles {
+		b, err := registry.NewImageInput(to, from)
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to parse unpacked bundle image %s", to))
+			continue
+		}
+
+		if _, ok := foundEntries[b.Bundle.Package]; !ok {
+			foundEntries[b.Bundle.Package] = map[string]map[registry.BundleKey]bool{}
+		}
+		v, err := b.Bundle.Version()
+		if err != nil {
+			errs = append(errs, fmt.Errorf("failed to parse version for %s (%s): %v", b.Bundle.Name, b.Bundle.BundleImage, err))
+		}
+		key := registry.BundleKey{
+			CsvName:    b.Bundle.Name,
+			Version:    v,
+			BundlePath: b.Bundle.BundleImage,
+		}
+		for _, c := range b.Bundle.Channels {
+			if _, ok := foundEntries[b.Bundle.Package][c]; !ok {
+				foundEntries[b.Bundle.Package][c] = map[registry.BundleKey]bool{}
+			}
+			foundEntries[b.Bundle.Package][c][key] = false
+		}
+	}
+
+	for pkg, channels := range foundEntries {
```

Comment:
same comment as above. if there's a reasonable way to decrease the overall complexity of this large function (i.e. fewer conditionals and nested blocks per function) by breaking it into smaller functions, it'll be easier to comprehend and maintain.